model.py

```diff
@@ -52,24 +52,28 @@ def __init__(self, config):
     def forward(self, x):
         B, T, C = x.size() # batch size, sequence length, embedding dimensionality (n_embd)
 
-        # calculate query, key, values for all heads in batch and move head forward to be the batch dim
-        q, k, v  = self.c_attn(x).split(self.n_embd, dim=2)
-        k = k.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
-        q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
-        v = v.view(B, T, self.n_head, C // self.n_head).transpose(1, 2) # (B, nh, T, hs)
+        # calculate query, key, values for all heads in batch
+        qkv = self.c_attn(x).view(B, T, 3, self.n_head, C // self.n_head)
 
         # causal self-attention; Self-attend: (B, nh, T, hs) x (B, nh, hs, T) -> (B, nh, T, T)
-        if self.flash:
+        if self.flash2:
             # efficient attention using Flash Attention CUDA kernels
-            y = torch.nn.functional.scaled_dot_product_attention(q, k, v, attn_mask=None, dropout_p=self.dropout if self.training else 0, is_causal=True)
+            from flash_attn import flash_attn_qkvpacked_func
+            y = flash_attn_qkvpacked_func(qkv, dropout_p=self.dropout if self.training else 0, causal=True)
+            # y: (B, T, nh, hs).
         else:
             # manual implementation of attention
+            qkv = qkv.transpose(1, 3) # (B, nh, 3, T, hs)
```

Comment:
for flash = False, a bunch more operations need to be done for every `forward` than before?